src/transform/src/join_implementation.rs

```diff
@@ -402,12 +419,21 @@ mod delta_queries {
             implementation,
         } = &mut new_join
         {
-            if inputs.len() < 2 {
+            if inputs.len() <= 2 {
+                // if inputs.len() == 0 then something is very wrong.
+                //
+                // if inputs.len() == 1:
                 // Single input joins are filters and should be planned as
                 // differential plans instead of delta queries. Because a
                 // a filter gets converted into a single input join only when
                 // there are existing arrangements, without this early return,
                 // filters will always be planned as delta queries.
+                // (ggevay: This is an old comment, and I'm not sure whether a single-input join
+                // could still actually occur. It is not happening in any of our slts currently.)
+                //
+                // if inputs.len() == 2:
+                // We decided to always plan this as a differential join for now, see here:
+                // https://github.com/MaterializeInc/materialize/pull/16099#issuecomment-1316857374
```

Comment:
> If we can differentiate between pre-existing arrangements and arranements resulting from a previous join implementation attempt, will this be sufficient to lift this restriction on delta joins?

Actually, the restriction of always planning 2-input joins as Delta is not a severe restriction. For 2-input joins, Differential and Delta have similar performance in _most_ cases, because the usual advantage of Delta join avoiding indexes on intermediate results doesn't apply, since there are not intermediate results at all.

But there are some cases when one is better than the other:
- Differential is sometimes better because the zigzagging can sometimes make the total work proportional to the output size instead of any input sizes, while a Delta join needs to always stream at least one input (more specifically, it streams one input in the initial snapshot; and then stream in diffs later for all inputs).
- Delta can be better [in some very special cases with an additional trick that our current join ordering code is currently not able to perform](https://github.com/MaterializeInc/materialize/pull/16099#issuecomment-1317035005).

(Also note that when planning a Differential join inserts enough `ArrangeBy`s to trick the next run into planning a Delta join, that is considered a _good thing_. In fact, I think this is the original reason for running JoinImplementation more than once. The reasoning is that in such a situation we are forced to create all the arrangements that a Delta join would need even if we ran a Differential join, so then why not make it a Delta join, and then we avoid the intermediate arrangements of a Differential join. The 2-input case is special, because there are no intermediate arrangements even for Differential join.)

> Can we remove the check from the plan method and change the end of action like this:

I moved it into the plan method, because I felt that the Delta join planning method should be the authority on whether we want a Delta join in the 2-input case. But I don't have a strong preference.

Reply:
> [...] because the usual advantage of Delta join avoiding indexes on intermediate results doesn't apply, since there are not intermediate results at all.

I forgot about that bit.

> Differential is sometimes better because the zigzagging can sometimes make the total work proportional to the output size instead of any input sizes

Maybe you can elaborate on that offline.